### PR TITLE
feat: Use the bloom filter to reduce ingester catalog queries, Luke

### LINKED ISSUE #8066: Reduce catalog queries in write path while both row & hash IDs are used
Since https://github.com/influxdata/influxdb_iox/pull/7963 merged, new partitions are addressed by deterministic hash-based IDs. Previously they were addressed by catalog row IDs.

For [reasons] we want to move queries out of the hot write path. This isn't possible when using row IDs, as we must find out if a partition exists or not before buffering data for it, as this ID is provided to the queriers for correlation purposes and is used for object store file path construction.

Until all old-style partitions age out, we must run in a "dual ID" mode, where new partitions use the hash-based IDs, and old partitions (prior to the above PR being merged) continue to use old-style row IDs.

During this time the queries must remain on the hot path, so we can do the existing "does it exist" check for old-style, row-ID-based partitions. However we can _practically_ eliminate queries for new partitions by using a probabilistic/space-efficient filter to only query for old partitions, plus some minor false-positive overhead. See https://github.com/influxdata/influxdb_iox/pull/7963#issuecomment-1602578784 for more.

[reasons]:https://docs.google.com/document/d/1YWjjnPPEdeTVX88nV5kvGRyRjXzSILjyWwa9GR9Z3bU/edit?usp=sharing

### PR DESCRIPTION
**Pull request by carols10cents**
~~THIS DEPENDS ON https://github.com/influxdata/influxdb_iox/pull/8475 MERGE THAT FIRST~~ Merged!

Closes #8066 

This PR:

- Wires up the `OldPartitionBloomFilter` so that when the ingester resolves the partition, it doesn't go to the ingester's partition cache or the catalog if it's definitely a new-style partition
- When the ingester fetches the sort key from the catalog for persistence, upsert the partition at that point rather than expecting it to exist
- Have end-to-end tests asking the catalog for partitions explicitly ask the ingester to persist before making the catalog request

I *think* all this is correct. Please review commit-by-commit!

### CODE REVIEW

**Comment by carols10cents** [comment]
Ok, so, problem. 

The ingester is no longer creating records in the catalog for new partitions for data coming in, which is what we want since we want fewer catalog interactions in the write path. Fine, fine.

The querier asks the ingesters for data and gets back some partitions if there are any:

https://github.com/influxdata/influxdb_iox/blob/81b5d80a91423f54f625a36f29b97b5f89d504aa/querier/src/table/mod.rs#L187-L193

Then it fetches those partitions from its partition cache:

https://github.com/influxdata/influxdb_iox/blob/81b5d80a91423f54f625a36f29b97b5f89d504aa/querier/src/table/mod.rs#L261-L268

By batching up `PartitionRequest`s from those partitions:

https://github.com/influxdata/influxdb_iox/blob/81b5d80a91423f54f625a36f29b97b5f89d504aa/querier/src/table/mod.rs#L351-L368

And `PartitionCache::get` only returns partitions that exist in its cache: 

https://github.com/influxdata/influxdb_iox/blob/81b5d80a91423f54f625a36f29b97b5f89d504aa/querier/src/cache/partition.rs#L158

but these partitions aren't going to exist in the querier's cache because they don't exist in the catalog yet.

This all looks like it has to do with sort keys and columns?

Hm, over here it says sort keys aren't needed for ingester partitions: https://github.com/influxdata/influxdb_iox/blob/81b5d80a91423f54f625a36f29b97b5f89d504aa/querier/src/table/mod.rs#L331-L335

And then over here where it's building the `CachedPartitions` it's building the columns from the table partition template and the partition key: https://github.com/influxdata/influxdb_iox/blob/81b5d80a91423f54f625a36f29b97b5f89d504aa/querier/src/cache/partition.rs#L235-L236

And we have the partition key in the ingester (we used it to create the `PartitionHashId`) but we aren't sending it directly to the querier, it has to fetch it from the catalog?

# TL;DR

So I'm not sure how to solve this problem of the querier assuming the partition will exist in its view of the catalog.

It seems like the querier might not neeeeeeed the partition record to exist in the catalog, if we can send all the info it needs from the ingester? Am I missing anything there?

Or should the ingester create the partition record in the catalog when it gets a query for that data? Does the querier refresh its cache in that case or does it only do that when there are new parquet files to pick up? 

Or should the querier be upserting partition records to the catalog when it hears about from the ingester?

HOW IS PARTITION RECORD FORMED NOW I DON'T KNOW

**Comment by crepererum** [comment]
So the `fetch_cached_partitions` result will not include the partition that is not in the catalog. However it will invalidate the cache for negative entries (= non-existing ones) for every request because non existing partitions don't contain the sort key.

Now what happens for these non existing partitions: they are only used for ingester partitions (otherwise they exist in the catalog and the cache invalidation would have made sure that we fetch them), and when we process them here:

https://github.com/influxdata/influxdb_iox/blob/81b5d80a91423f54f625a36f29b97b5f89d504aa/querier/src/table/mod.rs#L285-L287

**This code is wrong**, because it will drop ingester partitions if there is no cached partition. I think it should be instead: "if there's a cached partition, attach the column ranges, otherwise leave the partition as it is".

Now this results in a slightly less efficient pruning, but I think this is OK since in theory the ingester already pruned the partitions, so I think the end result is the same.

We could try to fix the cache invalidation spam, but I think this is good enough for now.

---

To come to your questions:

> It seems like the querier might not neeeeeeed the partition record to exist in the catalog, if we can send all the info it needs from the ingester? Am I missing anything there?

As written above, I think we don't need them AT ALL because the ingester already prunes the partition. The code bug above needs to be fixed though.

> Or should the ingester create the partition record in the catalog when it gets a query for that data?

Not required, see prev. point.

> Does the querier refresh its cache in that case or does it only do that when there are new parquet files to pick up?

As I wrote during the intro, the querier will invalidate always for negative entries. This is quite spammy and we could optimize that later.

> Or should the querier be upserting partition records to the catalog when it hears about from the ingester?

Not required, see first answered question.

> HOW IS PARTITION RECORD FORMED NOW I DON'T KNOW

What's a partition record? Are you talking about:

- the catalog
- the cache
- the ingester->querier wire format

I think I could answer this question if you specify "record". Or I could answer all 3 versions if you want.

**Comment by carols10cents** [comment]
> **This code is wrong**, because it will drop ingester partitions if there is no cached partition. I think it should be instead: "if there's a cached partition, attach the column ranges, otherwise leave the partition as it is".

Cool ok I can change that!


> What's a partition record? Are you talking about:
> 
>     * the catalog
> 
>     * the cache
> 
>     * the ingester->querier wire format
> 
> 
> I think I could answer this question if you specify "record". Or I could answer all 3 versions if you want.

I'm talking specifically about the catalog record here. If the ingester is no longer creating it, when *should* it be created? And is that happening now? I ran out of time yesterday and didn't get to figuring that out.

**Comment by carols10cents** [comment]
Ok, now I'm getting a panic because the chunk stats aren't set (because `set_partition_column_ranges` was doing that but I made the change to not call that if there's no cached partition :( https://github.com/influxdata/influxdb_iox/blob/614b64d3e416f81bc9841896a29fcff0762a3b40/querier/src/ingester/mod.rs#L929

**Comment by crepererum** [comment]
> f the ingester is no longer creating it, when _should_ it be created? And is that happening now?

The ingester creates it when it persists and this is sufficient for the querier (modulo the bug mentioned by me).

**Comment by crepererum** [comment]
> Ok, now I'm getting a panic because the chunk stats aren't set (because `set_partition_column_ranges` was doing that but I made the change to not call that if there's no cached partition :(

Then maybe call `set_partition_column_ranges` with empty ranges (IIRC this is a dict, should be implement `Default`).

**Comment by carols10cents** [comment]
@crepererum That works!!! I reverted my commit where I tried setting basic stats on ingester chunks; happy to rebase that out if you'd like.

**Comment by philjb** [comment]
the force is strong with this one

**Comment by crepererum** [comment]
> That works!!! I reverted my commit where I tried setting basic stats on ingester chunks; happy to rebase that out if you'd like.

Yeah, if you could extract the querier changes into a mini-PR, that would be great. The reason is that I would like to see this landed separately so if something explodes, we at least know which part of this PR was the cause (I don't expect the querier changes to do anything bad, but better be safe than sorry).

File: ingester/src/buffer_tree/partition/resolver/sort_key.rs [hunk 1]
```diff
@@ -34,14 +37,10 @@ impl SortKeyResolver {
         Backoff::new(&self.backoff_config)
             .retry_all_errors("fetch partition sort key", || async {
                 let mut repos = self.catalog.repositories().await;
-                let s = partition_lookup(repos.as_mut(), &self.partition_id)
+                let s = repos
+                    .partitions()
+                    .create_or_get(self.partition_key.clone(), self.table_id)
```
Comment (domodwyer) [inline ingester/src/buffer_tree/partition/resolver/sort_key.rs:42]: Hmm, I suspect us adding more things to the `Partition` is going to make this more complicated in the future, because we'll have to pass all the info needed to make a partition into the "get the partition" code. It changes the direction of data flow a fair bit!

Is this just for tests?

File: ingester/src/init.rs [hunk 2]
```diff
@@ -319,6 +334,14 @@ where
...
         Arc::clone(&metrics),
     );
+    let partition_provider = OldPartitionBloomFilter::new(
+        partition_provider,
+        Arc::clone(&catalog),
+        BackoffConfig::default(),
+        persist_background_fetch_time,
+        Arc::clone(&metrics),
+        old_style,
+    );
```
Comment (domodwyer) [inline ingester/src/init.rs:344]: I think we probably want to do have the call chain be:

```
         │
         │    ┌────────────────┐
         │    │     Cache      │
         │    └────────────────┘
         │    ┌────────────────┐
         │    │  Bloom Filter  │
         │    └────────────────┘
      Get()   ┌────────────────┐
         │    │    Coalesce    │
         │    └────────────────┘
         │    ┌────────────────┐
         │    │    Catalog     │
         │    └────────────────┘
         │
         ▼
```

That way we use the cache first (100% accuracy!) and slowly drain it (it removes cache hits from the cache, as they'll only be queried for once) to recover the memory it uses.

Although the 🐹 is napping, so maybe I've forgotten something!

File: iox_catalog/src/postgres.rs [hunk 3]
```diff
@@ -1499,6 +1499,20 @@ LIMIT $1;"#,
             .await
             .map_err(|e| Error::SqlxError { source: e })
     }
+
+    async fn list_old_style(&mut self) -> Result<Vec<Partition>> {
```
Comment (domodwyer) [inline iox_catalog/src/postgres.rs:1503]: Let save ourselves from ourselves:

```suggestion
    async fn list_old_style(&mut self) -> Result<Vec<Partition>> {
        // Correctness: the main caller of this function, the partition bloom
        // filter, relies on all partitions being made available to it.
        //
        // This function MUST return the full set of old partitions to the
        // caller - do NOT apply a LIMIT to this query. 
        // 
        // The load this query saves vastly outsizes the load this query causes.
```

**Review by domodwyer — COMMENTED** [review]
Looking good! Bit unsure about the change in data flow - I think it's OK for now if it's needed, but ideally we don't make the getter have to know everything to construct the actual partition, as the latter is likely to get more complex over time.

File: ingester/src/buffer_tree/partition/resolver/sort_key.rs (hunk 1 above)
Comment (carols10cents) [reply ingester/src/buffer_tree/partition/resolver/sort_key.rs:42]: This is now where partitions get created if they're not created already, and it is kind of by coincidence, that the first place we need to fetch the sort key is the first place that we really need the partition record to exist in the catalog for persistence. 

I'll try rearranging and see if it's straightforward enough to include in this PR.

File: ingester/src/init.rs (hunk 2 above)
Comment (carols10cents) [reply ingester/src/init.rs:344]: I'll give that a try too!

Comment (carols10cents) [reply ingester/src/init.rs]: Rearranged and this seems to work!

File: ingester/src/buffer_tree/partition/resolver/sort_key.rs (hunk 1 above)
Comment (carols10cents) [reply ingester/src/buffer_tree/partition/resolver/sort_key.rs]: Ok, I'd like you to take a look at the changes I made before I merge this. I reverted the commit that made this change (will rebase that commit+revert out before merging), and added a call to a function that calls `PartitionRepo::create_or_get` in the persist worker's `compact_and_upload`. Does that seem like a better place/way to create the partition record if needed? I don't feel sure.

File: ingester/src/buffer_tree/partition/resolver/sort_key.rs [hunk 4]
```diff
@@ -36,13 +36,7 @@ impl SortKeyResolver {
                 let mut repos = self.catalog.repositories().await;
                 let s = partition_lookup(repos.as_mut(), &self.partition_id)
                     .await?
-                    .unwrap_or_else(|| {
-                        panic!(
-                            "resolving sort key for non-existent partition ID {}",
-                            self.partition_id
-                        )
-                    })
-                    .sort_key();
```
Comment (domodwyer) [inline ingester/src/buffer_tree/partition/resolver/sort_key.rs:45]: Ahhh bananas.

So the problem with this is that the following happens:

* Write comes in for a new partition that does not exist
* It's a new partition, so it's a hash-addressed partition, catalog record not created
* We initialise the `SortKeyResolver` and process the write
* Some time later, the deferred `SortKeyResolver` fires, and we query the catalog for the partition sort key
* The partition doesn't exist! (this is the panic removed here)
* Retry the query
* The partition doesn't exist!
* Retry the query
* The partition doesn't...

So we'll wind up firing loads of queries at the catalog _until_ the partition record is created, for every new partition. 

Worse still, if the code is structured so that the sort key is awaited before the record creation call, it'll deadlock, because it'll be waiting for a partition to be created before it creates the partition :fire:

We could just delay the deferred sort key resolver to be "infinitely in the future" - for row-addressed, use the normal deferred timeout, and for hash-addressed, use `Duration::MAX` or something. We'd have to be careful / have a test that ensures the deadlock above doesn't happen too.

Or we can back it out, and pretend we never went down this path - sorry for the messing around - I didn't foresee this 🐹

**Review by domodwyer — COMMENTED** [review]
One big party pooper 💩 Otherwise it looks excellent.

Perhaps we back the "information direction" change out, and get this merged. We can come back to the "where information flows" problem later, if it becomes a serious problem. Sorry for the detour - I thought it'd be an easy tweak.

**Review by domodwyer — APPROVED** [review]
I'm going to approve this so you can get it merged after backing out the misguided change I totally goose-chased you on :+1:

**Comment by carols10cents** [comment]
Ok, I put the sort key resolver code back to the way it was when you first reviewed, and updated the PR description accordingly. Here goes!
